File: algo_2021/s6.-dfs-comb-perm/15.Permutations/15_permutation.py
class Solution:
    """
    @param: nums: A list of integers.
    @return: A list of permutations.
    """
    def permute(self, nums):
        # write your code here
        res = []
        if nums == None: 
            return res
        if len(nums) == 0:
            res.append(list(res))
            return res
        self.dfs(0, nums, set(), [], res)
        return res
    
    def dfs(self, index, nums, visited, permutation, res):
        if len(visited) == len(nums):
            # WARNING!
            # should deep copy list since list is pass by reference 
            res.append(list(permutation))
            return
        for i in range(len(nums)):
            if i in visited:
                continue
            # add to already visited 
            visited.add(i)
            permutation.append(nums[i])
            self.dfs(i, nums, visited, permutation, res)
            # backtracing, remove top element
            permutation.pop()
            # remove already visited record
            visited.remove(i)

File: lintcode_algo_2021/s6.-dfs-comb-perm/15.Permutations/15_permutation.java
class Solution {
    /*
     * @param nums: A list of integers.
     * @return: A list of permutations.
     */
    public List<List<Integer>> permute(int[] nums) {
        // write your code here
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null) {
            return res;
        }

        if (nums.length == 0) {
            res.add(new ArrayList<Integer>());
            return res;
        }

        boolean[] visited = new boolean[nums.length];
        dfs(nums, visited, new ArrayList<Integer>(), res);
        return res;
    }

    private void dfs(int[] nums,
                     boolean[] visited, 
                     List<Integer> permutation,
                     List<List<Integer>> res
                    ) {
        if (permutation.size() == nums.length) {
            res.add(new ArrayList<Integer>(permutation));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (visited[i]) {
                continue;
            }
            visited[i] = true;
            permutation.add(nums[i]);
            dfs(nums, visited, permutation, res);
            permutation.remove(permutation.size() - 1);
            visited[i] = false;
        }